### ISSUE TITLE
decode_loop: pipelined_decode_stops_on_eos flakes in full suite (test isolation / MLX global state)

### ISSUE THREAD

**Issue by Pushkinist** [Pushkinist/rMLX#100]
## Summary

`decode_loop::decode_loop_tests::pipelined_decode_stops_on_eos` passes in isolation and when its own module is run alone, but **intermittently fails** when the full crate test suite runs in parallel (`cargo test -p rmlx-models --release --lib`, and via `make ci`'s workspace test).

## Symptom

The emitted token-id stream is polluted with garbage values instead of the scripted `[0, 1, 3]`:

```
assertion `left == right` failed: decode stops once EOS is emitted
  left: [0, 897988541, 1056964608, 1015021568, 897988541, 1057133550, 0, 1072857967, ... , 1, 1, 3]
 right: [0, 1, 3]
```

The garbage values are f32 bit patterns reinterpreted as u32 (e.g. `1065353216` = `1.0f32`, `1056964608` = `0.5f32`, `3212836864` = `-1.0f32`), i.e. the pipelined decode step is reading leftover / concurrently-mutated MLX array state rather than the scripted logits.

## Reproduction

```
cargo test -p rmlx-models --release --lib        # run a few times — intermittent FAIL
cargo test -p rmlx-models --release --lib decode_loop::decode_loop_tests::pipelined_decode_stops_on_eos -- --exact   # always PASS
```

Passes isolated and module-only; fails only under full-suite parallelism → classic test-isolation / shared-global-state bug.

## Root-cause hypothesis (unconfirmed)

The test acquires the global `mlx_guard()` MLX serialization lock. The corruption suggests another MLX-/`Array`-touching test runs **concurrently without acquiring that guard** (or the pipelined decode path performs work that escapes the guard's scope), so Metal/MLX process-global state is mutated underneath this test.

## Candidate directions

- Audit `rmlx-models` tests that construct `Array`s / touch MLX without `mlx_guard()` and add the guard.
- Or fix the pipelined-decode path's test so its MLX work is fully bracketed by the guard.
- Prefer a correct isolation fix over a blanket `--test-threads=1` workaround.

Found while running the pre-merge gate on an unrelated branch.
